diff --git a/apps/system/js/activity_window.js b/apps/system/js/activity_window.js
index 6d41b42..141a6d4 100644
--- a/apps/system/js/activity_window.js
+++ b/apps/system/js/activity_window.js
@@ -149,7 +149,9 @@
 
   ActivityWindow.prototype._transitionHandler =
     function acw__transitionHandler(evt) {
-      evt.stopPropagation();
+      if (evt) {
+        evt.stopPropagation();
+      }
       if (this.element.classList.contains('opening')) {
         this.element.classList.remove('opening');
         this.element.classList.remove('slideup');
@@ -164,7 +166,7 @@
         }
         if (this.openCallback)
           this.openCallback();
-      } else {
+      } else if (this.element.classList.contains('closing')) {
         this.element.classList.remove('closing');
         this.element.classList.remove('slidedown');
         this.element.classList.remove('active');
@@ -197,7 +199,7 @@
         }
         this.debug('request caller to open again');
         if (this.activityCaller instanceof ActivityWindow) {
-          if (evt) {
+          if (evt && this.activityCaller.isActive()) {
             this.activityCaller.open();
           }
         } else if (this.activityCaller instanceof AppWindow) {
@@ -300,6 +302,7 @@
     this.restoreCaller();
     this.element.classList.add('slidedown');
     this.element.classList.add('closing');
+    setTimeout(this._transitionHandler.bind(this));
   };
 
   ActivityWindow.prototype.show = function acw_show() {
diff --git a/apps/system/js/activity_window_factory.js b/apps/system/js/activity_window_factory.js
index 6666212..f18fcc6 100644
--- a/apps/system/js/activity_window_factory.js
+++ b/apps/system/js/activity_window_factory.js
@@ -154,7 +154,8 @@
         case 'activityterminated':
           this._activities.forEach(function iterator(activity, index) {
             if (activity.instanceID === evt.detail.instanceID) {
-              if (activity.instanceID === this._lastActivity.instanceID) {
+              if (this._lastActivity &&
+                  activity.instanceID === this._lastActivity.instanceID) {
                 this._lastActivity = null;
               }
               if (this._activeActivity &&
@@ -188,7 +189,8 @@
           this.debug('activity: ' + activity.name +
             ' is closing, its caller is ' + activity.activityCaller.name);
           if (activity.activityCaller &&
-              activity.activityCaller instanceof ActivityWindow) {
+              activity.activityCaller instanceof ActivityWindow &&
+              activity.activityCaller.isActive()) {
             this._activeActivity = activity.activityCaller;
           } else if (this._activeActivity &&
               this._activeActivity.instanceID == evt.detail.instanceID) {
