diff --git a/b2g/chrome/content/shell.js b/b2g/chrome/content/shell.js
index 4529b31..424b7a7 100644
--- a/b2g/chrome/content/shell.js
+++ b/b2g/chrome/content/shell.js
@@ -1033,7 +1033,12 @@ let RemoteDebugger = {
   _running: false,
 
   prompt: function debugger_prompt() {
-    this._promptDone = false;
+  try {
+ //     if (Services.prefs.getBoolPref("debugger.bypass_prompt") === true) {
+        return true;
+   //   }
+   } catch(e) {}  
+  this._promptDone = false;
 
     shell.sendChromeEvent({
       "type": "remote-debugger-prompt"
